# Week 2 - Mini Assessment

Welcome to your mini assessment for week two. A set of exercises using NumPy to make sure you get accustomed to one of the most widely used libraries in the data analytics/science community.

Once you are done with the notebook, you should try to put it on GitHub under your respective folder for everything related to module 2. In addition let your instructor know that you have submitted this mini-assessment.

1. Import the numpy package and check its version.

In [1]:
#1
import numpy as np

2. Create an numpy range of values from 0 to 49 and name it `mynums`. Print its output.

In [2]:
#2
mynums=np.arange(50)
mynums

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
       34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])

3. Reshape the array you created in step 2 into a matrix with the shape below and assign a copy of it to a variable called `my_matrix`. Print the matrix.

In [33]:
#3
my_matrix=mynums.reshape(5,10)
print(my_matrix, '\n')
print('My_matrix shape is:',my_matrix.shape)

[[ 0  1  2  3  4  5  6  7  8  9]
 [10 11 12 13 14 15 16 17 18 19]
 [20 21 22 23 24 25 26 27 28 29]
 [30 31 32 33 34 35 36 37 38 39]
 [40 41 42 43 44 45 46 47 48 49]] 

My_matrix shape is: (5, 10)


4. Create a linearly-spaced array of size 100 between the numbers 20 to 100, assign it to the variable, `num_increases` and print it.

In [4]:
#4
num_increase=np.linspace(20,100,100)
num_increase

array([ 20.        ,  20.80808081,  21.61616162,  22.42424242,
        23.23232323,  24.04040404,  24.84848485,  25.65656566,
        26.46464646,  27.27272727,  28.08080808,  28.88888889,
        29.6969697 ,  30.50505051,  31.31313131,  32.12121212,
        32.92929293,  33.73737374,  34.54545455,  35.35353535,
        36.16161616,  36.96969697,  37.77777778,  38.58585859,
        39.39393939,  40.2020202 ,  41.01010101,  41.81818182,
        42.62626263,  43.43434343,  44.24242424,  45.05050505,
        45.85858586,  46.66666667,  47.47474747,  48.28282828,
        49.09090909,  49.8989899 ,  50.70707071,  51.51515152,
        52.32323232,  53.13131313,  53.93939394,  54.74747475,
        55.55555556,  56.36363636,  57.17171717,  57.97979798,
        58.78787879,  59.5959596 ,  60.4040404 ,  61.21212121,
        62.02020202,  62.82828283,  63.63636364,  64.44444444,
        65.25252525,  66.06060606,  66.86868687,  67.67676768,
        68.48484848,  69.29292929,  70.1010101 ,  70.90

5. Create an numpy array with 20 zeros, assign it to the variable `zeroes`.

In [34]:
#5
zeroes=np.zeros(20, dtype="int8")
zeroes

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
      dtype=int8)

6. Create a two dimensional array of size 7 by 8 filled with 1s and do the following:
<br>-6.1 mulptiply the first rows by 2
<br>-6.2 add 3, and raise to the power of two the odd columns starting from the beginning
<br>-6.3 subtract 10 and multiply by 15 the elements in the even rows and odd columns

In [70]:
#6.Create a two dimensional array of size 7 by 8 filled with 1s and do the following:
ones=np.ones((7,8), dtype='int8')
print(ones)
print()
print('The array dimension is:', ones.ndim )

[[1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]
 [1 1 1 1 1 1 1 1]]

The array dimension is: 2


In [71]:
#6.1. mulptiply the first rows by 2

ones[0]=ones[0]*2
ones[0]

array([2, 2, 2, 2, 2, 2, 2, 2], dtype=int8)

In [73]:
#6.2. add 3, and raise to the power of two the odd columns starting from the beginning
odd_col_index=[1,3,5,7]

ones[:,1::2]=(ones[:,1::2]+3)**2
ones[:,1::2]

array([[25, 25, 25, 25],
       [16, 16, 16, 16],
       [16, 16, 16, 16],
       [16, 16, 16, 16],
       [16, 16, 16, 16],
       [16, 16, 16, 16],
       [16, 16, 16, 16]], dtype=int8)

In [74]:
#6.3. subtract 10 and multiply by 15 the elements in the even rows and odd columns
ones[::2,1::2]=(ones[::2,1::2]-10)*15
ones[::2,1::2]

(7, 8)


array([[-31, -31, -31, -31],
       [ 90,  90,  90,  90],
       [ 90,  90,  90,  90],
       [ 90,  90,  90,  90]], dtype=int8)

7. Create a numpy array starting at 20 and ending at 40 with 10 equally spaced values. Print your result.

In [12]:
np.linspace(20,40,10)

array([20.        , 22.22222222, 24.44444444, 26.66666667, 28.88888889,
       31.11111111, 33.33333333, 35.55555556, 37.77777778, 40.        ])

8. Create a 9 by 9 matrix with random integers from 100 to 500 and use the dtype "int32".

In [89]:
print("\nBelow is the initial matrix:\n")
matrix=np.random.randint(100,500, size=(9*9),dtype='int32')

print(matrix)
print("\nBelow is the reshaped matrix")
matrix.reshape(9,9)



Below is the initial matrix:

[365 131 444 138 239 400 271 363 212 196 449 460 252 399 279 366 194 149
 374 224 238 161 490 235 307 496 471 228 359 151 499 295 181 415 422 499
 417 119 272 443 161 121 476 484 419 194 229 440 334 288 304 122 242 274
 336 204 388 308 298 410 252 317 263 157 259 350 395 370 469 210 369 119
 243 453 269 263 125 481 279 134 250]

Below is the reshaped matrix


array([[365, 131, 444, 138, 239, 400, 271, 363, 212],
       [196, 449, 460, 252, 399, 279, 366, 194, 149],
       [374, 224, 238, 161, 490, 235, 307, 496, 471],
       [228, 359, 151, 499, 295, 181, 415, 422, 499],
       [417, 119, 272, 443, 161, 121, 476, 484, 419],
       [194, 229, 440, 334, 288, 304, 122, 242, 274],
       [336, 204, 388, 308, 298, 410, 252, 317, 263],
       [157, 259, 350, 395, 370, 469, 210, 369, 119],
       [243, 453, 269, 263, 125, 481, 279, 134, 250]])

9. Create an empty array of 20 elements with shape 2 by 10, dtype 'float32' and add a different number to each row.

In [113]:
empty_array=np.empty((2, 10),dtype='float32')
print(empty_array)

empty_array+5

[[11.404966  10.327561  10.290468  11.166316  11.088625  10.032078
  10.95538   10.621864  10.336212  11.894518 ]
 [10.57457   11.014786  10.55241   10.389251  10.497087  10.685261
  11.269937  11.156866  10.155604  11.3226385]]


array([[16.404966, 15.327561, 15.290468, 16.166317, 16.088625, 15.032078,
        15.95538 , 15.621864, 15.336212, 16.894518],
       [15.57457 , 16.014786, 15.55241 , 15.389251, 15.497087, 15.685261,
        16.269936, 16.156866, 15.155604, 16.32264 ]], dtype=float32)

10. Create an array with 15 random numbers between 100 and 200 and print the maximum value of the array.

In [15]:
np.random.seed(1)
a=np.random.randint(100,200,15)
print(a)
print("The max value of the array is:",max(a))
#double check if this is the correct answer for Q10.

[137 112 172 109 175 105 179 164 116 101 176 171 106 125 150]
The max value of the array is: 179


11. Create an array with the following numbers [40, 41, 26, 14, 15, 47, 43, 33, 44, 22], and assign it to a variable
<br>- 11.1 create a slice from the 5th to 10th element, and assign it to a variable </b>
<br>- 11.2 change the value of both arrays with any operation you'd like (add, subtract, multiply, etc.), and </b>
<br>- 11.3 then print all 3 (the first array and your next two variables).</b>

In [114]:
#11.Create an array with the following numbers [40, 41, 26, 14, 15, 47, 43, 33, 44, 22], and assign it to a variable
var1=np.array([40, 41, 26, 14, 15, 47, 43, 33, 44, 22])
var1

array([40, 41, 26, 14, 15, 47, 43, 33, 44, 22])

In [115]:
#11.1 create a slice from the 5th to 10th element, and assign it to a variable
var2=var[4:10]
var2

array([15, 47, 43, 33, 44, 22])

In [116]:
#11.2 change the value of both arrays with any operation you'd like (add, subtract, multiply, etc.), and
var2_new=var2*10
var1_new=var1-3

In [117]:
print('var1_new',var1_new)
print('var2_new',var2_new)


var1_new [37 38 23 11 12 44 40 30 41 19]
var2_new [150 470 430 330 440 220]


12. Create an array of 360 random numbers from -10 to 50 (to resemble the weather) and find the min, max, 25th percentile, 75 percentile, average, and the median.

In [219]:
np.random.seed(2)
temp=np.random.randint(-10,50,360)
temp1=temp.reshape(1,360)
temp

array([ 30,   5,  35,  -2,  12,  33,   8,   1,  30,  -3,  24,  39,  21,
         1,  11,  37,  21,  16,  10,  42,  27,  29,  -7,  28,  -6,  32,
        33,  41,  29,  28,  32,  23,  48,  -7,  -5,  14,  -6,  36,  -4,
        21,   9,  47,  21,  -8,   6,  42,  36,   2,  40,  -6,  16,   5,
        39,  29,  36,  -2,  40,  35,  48,   5,  31,  35,  -2,  45,   7,
        12,  46,  48,  43,  -1,  47,  31,  36,  16,   9,  22,  33,  22,
        16,  40,  -2,   2,   0,  30,  24,  -1,  44,  27,  -4,  12,  -4,
         9,  46,   8,  -9,  -6,  30,   7,  -4,  47,  27,  23,   8,  10,
        16,  13,  12,  33,  42,  27,   0,  -2,  16,  25,  17,  22,   6,
        43,  45,  49,  11,  33,  19,   6,  30,  -4,   0,  27,  49,   7,
         5,  20,   3,  16,  29,  33,  37,  10,  41,  12,  49,  39,  26,
        48,  17, -10,  30,  48,  43,  18,  26,   1,  36,  40,  -1,  -3,
        39,  47,   8,  -9,   5,  17,  36,  22,  29,  39,  30,  34,  36,
        49,  -1,  33,   4,  20,  36,  21,  22,  45,   6,  11,  3

In [231]:
#find the min, max, 25th percentile, 75 percentile, average, and the median.
def temp_stats(temp):
    print("Temp Min:", np.min(temp))
    print("Temp Max:", np.max(temp))
    print("Temp 25th percentile:", np.percentile(temp,25))
    print("Temp 75th percentile:", np.percentile(temp,75))
    print("Temp avg:", round(np.mean(temp),1))
    print("Temp median:", np.median(temp))
    
    
temp_stats(temp=temp)

Temp Min: -10
Temp Max: 49
Temp 25th percentile: 7.0
Temp 75th percentile: 35.0
Temp avg: 20.6
Temp median: 21.0


13. Reshape the array you created in step 12 into a matrix of 300 by 12 and get the same values but only for the months this time.

In [232]:
temp2=temp1.reshape((12,30))
print("new shape is:",temp2.shape)

new shape is: (12, 30)


In [236]:
months=["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"]

for x in range(len(temp2)):
    month_num=months[x]
    month_temp=temp2[x]
    
    print(f"Temp Stats - {month_num}:")
    temp_stats(temp=month_temp)
    print()    

Temp Stats - Jan:
Temp Min: -7
Temp Max: 42
Temp 25th percentile: 8.5
Temp 75th percentile: 31.5
Temp avg: 20.2
Temp median: 25.5

Temp Stats - Feb:
Temp Min: -8
Temp Max: 48
Temp 25th percentile: 2.75
Temp 75th percentile: 36.0
Temp avg: 19.7
Temp median: 21.0

Temp Stats - Mar:
Temp Min: -4
Temp Max: 48
Temp 25th percentile: 7.5
Temp 75th percentile: 35.75
Temp avg: 22.3
Temp median: 23.0

Temp Stats - Apr:
Temp Min: -9
Temp Max: 49
Temp 25th percentile: 7.25
Temp 75th percentile: 29.25
Temp avg: 18.5
Temp median: 16.0

Temp Stats - May:
Temp Min: -10
Temp Max: 49
Temp 25th percentile: 10.25
Temp 75th percentile: 36.0
Temp avg: 23.1
Temp median: 23.0

Temp Stats - Jun:
Temp Min: -9
Temp Max: 49
Temp 25th percentile: 6.5
Temp 75th percentile: 36.0
Temp avg: 22.6
Temp median: 24.0

Temp Stats - Jul:
Temp Min: -9
Temp Max: 49
Temp 25th percentile: 7.5
Temp 75th percentile: 33.0
Temp avg: 21.6
Temp median: 22.0

Temp Stats - Aug:
Temp Min: -2
Temp Max: 49
Temp 25th percentile: 9.0
Temp 7

14. Create a mask for your data, and filter out and print:
<br>-14.1 the average temperature of the days where the temperature was below 10
<br>-14.2 the average temperature of the days where the temperature was greater than 30

In [164]:
#14.1 the average temperature of the days where the temperature was below 10
# temp2[temp2<10]
temp_under_10=temp2[temp2<10]
avg_temp=round(np.mean(temp_under_10),2)
print('Avg temp is: ',avg_temp)
print('\nAll temps that is under 10:', temp_under_10)

Avg temp is:  -0.28

All temps that is under 10: [  2  -2  -1   1  -5   5 -10   6  -9   2  -3  -4   8   1   4  -6   3  -1
  -3  -9 -10   7  -2   3  -2  -3  -7  -4  -7  -6   2   6   8   5 -10  -6
  -3  -1  -7   9  -2   0   5  -3   0  -9   3  -7 -10   3  -4  -4  -8   2
   1  -3   3  -2   1   2  -3  -6   8   3   0   7  -3   0   2  -9   8  -8
   0   8  -4  -8  -3   5  -5   7   3   0   9   4   3  -4   3   5  -1  -8
  -3  -5   3   3   4   6  -7  -8   4 -10   1  -8   9   8   0  -2  -2   3
   5 -10   9   5   4   9  -3   9  -7   9  -5   5   0  -8   7  -6 -10   4
   7   7   0   3  -4 -10   6   9  -1]


In [238]:
#14.2  the average temperature of the days where the temperature was greater than 30
temp_above_30=temp2[temp2>30]
avg_temp=round(np.mean(temp_above_30),2)
print('Avg temp is: ',avg_temp)
print('\nAll temps that is above 30:', temp_above_30)

Avg temp is:  40.17

All temps that is above 30: [35 33 39 37 42 32 33 41 32 48 36 47 42 36 40 39 36 40 35 48 31 35 45 46
 48 43 47 31 36 33 40 44 46 47 33 42 43 45 49 33 49 33 37 41 49 39 48 48
 43 36 40 39 47 36 39 34 36 49 33 36 45 33 48 32 39 41 33 37 33 49 31 48
 47 32 38 35 40 49 46 43 39 41 41 39 34 42 46 39 35 33 47 39 31 31 44 48
 49 37 42 41 35 43 41 32 35 49 47 43 48 31 48 48 41 37 44]


15. Create a full 10 by 10 matrix with a number of your choosing and assign it to a variable. After printing that variables, unravel the matrix and assign the array to a new number.

In [166]:
#create the matrix
maxtrix= np.random.randint(5,125,(10,10))
maxtrix

array([[ 56, 116,  15,  73,  28,  19,  68,  26,  51,   8],
       [ 61,  93,  85,  51,  59,  84,  76,  19,  82,  20],
       [ 30,  58,  89,  63,  90, 119, 115,  34,  49,  42],
       [ 27,  96,  59,  81,  17,  64,  31,  81,  76,  92],
       [ 44,  48,  81, 119,  43,  96, 117,  74, 103,  38],
       [ 48,  31,  61,  74,  78,  57,  94,  32,  48,   6],
       [ 87,  99, 101,  31,  89,  35, 108,  69,  27, 103],
       [ 57,   8,  75,  17,  44, 107,  88, 116,  53,  66],
       [ 75,  18,  41,  28,  27,  71,  97, 111, 100,  58],
       [ 14,  46,  62,  93, 114, 123,  98, 121,  81,  87]])

In [167]:
#unravel the matrix
unravelled_maxtrix=maxtrix.ravel()
unravelled_maxtrix

array([ 56, 116,  15,  73,  28,  19,  68,  26,  51,   8,  61,  93,  85,
        51,  59,  84,  76,  19,  82,  20,  30,  58,  89,  63,  90, 119,
       115,  34,  49,  42,  27,  96,  59,  81,  17,  64,  31,  81,  76,
        92,  44,  48,  81, 119,  43,  96, 117,  74, 103,  38,  48,  31,
        61,  74,  78,  57,  94,  32,  48,   6,  87,  99, 101,  31,  89,
        35, 108,  69,  27, 103,  57,   8,  75,  17,  44, 107,  88, 116,
        53,  66,  75,  18,  41,  28,  27,  71,  97, 111, 100,  58,  14,
        46,  62,  93, 114, 123,  98, 121,  81,  87])

In [168]:
#assign the array to the new value
new_array_value=(unravelled_maxtrix)*3-20
new_array_value

array([148, 328,  25, 199,  64,  37, 184,  58, 133,   4, 163, 259, 235,
       133, 157, 232, 208,  37, 226,  40,  70, 154, 247, 169, 250, 337,
       325,  82, 127, 106,  61, 268, 157, 223,  31, 172,  73, 223, 208,
       256, 112, 124, 223, 337, 109, 268, 331, 202, 289,  94, 124,  73,
       163, 202, 214, 151, 262,  76, 124,  -2, 241, 277, 283,  73, 247,
        85, 304, 187,  61, 289, 151,   4, 205,  31, 112, 301, 244, 328,
       139, 178, 205,  34, 103,  64,  61, 193, 271, 313, 280, 154,  22,
       118, 166, 259, 322, 349, 274, 343, 223, 241])